decimal/decimal.go

```diff
@@ -0,0 +1,158 @@
+package decimal
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+	"strings"
+)
+
+// Decimal represents a decimal amount, internally stored as a big.Int.
+type Decimal struct {
+	i     *big.Int
+	scale int
+}
+
+// New returns a new Decimal. scale specifies the units of i as an inverse power
+// of 10. For example, a scale of 2 indicates that i is in units of 10^-2, and
+// a value of i=25 would represent 25*10^-2=0.25. If scale=0, then the decimal
+// is an integer.
+func New(i *big.Int, scale int) Decimal {
+	return Decimal{
+		i:     i,
+		scale: scale,
+	}
+}
+
+// MarshalJSON converts the Decimal to JSON bytes.
+func (d Decimal) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + d.String() + `"`), nil
+}
+
+// UnmarshalJSON reads JSON bytes into a Decimal.
+func (d *Decimal) UnmarshalJSON(b []byte) error {
+	s := strings.Trim(string(b), `"`)
+	if strings.IndexByte(s, 'e') > -1 {
+		return errors.New("luno: unsupported decimal notation")
+	}
+	f, err := strconv.ParseFloat(s, 64)
+	if err != nil {
+		return err
+	}
+	scale := getScale(s)
+	i := big.NewInt(int64(f * math.Pow(float64(10), float64(scale))))
+	*d = New(i, scale)
+	return nil
+}
+
+// Strings returns a string representation of d.
+func (d Decimal) String() string {
+	di := bigIntDefault(d.i)
+	s := di.String()
+
+	var sign string
+	if di.Sign() < 0 {
+		sign = "-"
+		s = strings.TrimPrefix(s, "-")
+	}
+
+	if d.scale > 0 {
+		scale := int(d.scale)
+		if len(s) < scale+1 {
+			pad := scale + 1 - len(s)
+			s = strings.Repeat("0", pad) + s
+		}
+		return fmt.Sprintf("%s%s.%s", sign, s[:len(s)-scale], s[len(s)-scale:])
+	} else if d.scale < 0 {
+		return sign + s + strings.Repeat("0", -int(d.scale))
+	}
+
+	return sign + s
+}
+
+// Float64 converts the decimal d to a float64.
+func (d Decimal) Float64() float64 {
+	return float64(d.i.Int64()) * math.Pow(float64(10), float64(-d.scale))
+}
+
+// ToScale returns a Decimal representing the same value as d, but with the
+// given scale. d is left unchanged.
```

Comment:
Comment about what happens if the new scale is less than the existing scale. If it will round, is it toward zero (truncate towards zero)?